wikimedia/wikipedia-ios
# Implement all notification cell types

**Phabricator:** https://phabricator.wikimedia.org/T288658

### Notes
Adds Notifications Center cell and designs for our notification types.

- Implement the design of a notifications cell
- Ensure each of our notification types can be represented in a notifications cell
- Ensure the cell has two display presentations (read, unread) in three display states (default, edit selected, edit unselected) that match designs
- Ensure cell's leading image container can transform into a check-style selection control
- Add View Model (that will [later be populated with data](https://phabricator.wikimedia.org/T288662) from the Remote Notifications Controller) that represents all the notifications content for display in a cell

As I mentioned before, this PR is rather large. The commit history may help a bit, but if it feels easier to review as a sequence of smaller PRs, let me know and I'm certainly happy to break it up as so!

### Test Steps

1. Log in with an account that has notifications
2. Confirm notification cells appear in Notifications Center (ignore the actual content/data/images of the notifications, that's covered in a future Phab task)
3. Confirm notifications support the cell states below (using Figma as our visual guide)

#### Cell States Configuration

Unfortunately, the current steps to test the different presentation states of the cells is quite manual at the moment. It would benefit us to eventually have a test database of `RemoteNotification`s that include a notification of every type with every reasonable configuration. 

- **Notification Type**: `RemoteNotificationType` (thanks Toni) represents our available notification type. Update the type returned by `NotificationsCenterCellViewModel`'s `var notificationType: RemoteNotificationType?` to see each type, the image for that type, and the colors and visually match with Figma. This is not exhaustive.
- **Read/Unread**: cells can be presented as either read or unread. The color and font weights alter depending on this state. Update the Display State used within `NotificationsCenterCell`'s `func updateCellStyle(forDisplayState displayState: NotificationsCenterCellDisplayState)` to see the range of available presentations.
- **Selected/Unselected**: as well as being presented as read or unread, a cell can also be in edit mode as either selected or unselected. In edit mode, the leading image transforms from a notification type image to a selected or unselected checkmark. Update the same Display State above to see these presentations.
